### PR TITLE
Adding KafkaToGCS template and generalized  KafkaReader

### PR DESCRIPTION
**Pull request by nikhil6790**

### CODE REVIEW

**Review by vanshaj-google — APPROVED** [review]
Looks good!

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+          .format(gcsOutputFormat)
+          .outputMode(kafkaOutputMode)
+          .option("checkpointLocation", gcsCheckpointLocation)
+          .option("path", gcsOutputLocation)
+          .start()
+          .awaitTermination(kafkaAwaitTerminationTimeout);
+
+      LOGGER.info("KakfaToGCS job completed.");
+      spark.stop();
+    } catch (Throwable th) {
```
Comment (shashank-google) [inline java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java:138]: no try catch please.

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        KAFKA_TOPIC,
+        kafkaTopic,
+        KAFKA_STARTING_OFFSET,
+        kafkaStartingOffsets,
+        KAFKA_GCS_OUTPUT_MODE,
+        kafkaOutputMode,
+        KAFKA_GCS_AWAIT_TERMINATION_TIMEOUT,
+        kafkaAwaitTerminationTimeout);
+
+    try {
```
Comment (shashank-google) [inline java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java:113]: no try catch

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md [hunk 3]
```diff
@@ -125,4 +125,48 @@ bin/start.sh \
...
+kafka.schema.url=<gcs url for schema file>
+
+#Offset to start reading from. Values can be '**earliest**' ,'latest', '<topic_name>:{<partition>:<offset>}' ...
+kafka.starting.offset=<starting offset value>
+
+#Await time in milliseconds before terminating stream read
+kafka.gcs.await.termination.timeout=<stream await termination timeout>
+
+#Ouptut mode for writing data. Values can be 'append', 'complete', 'update'
+kafka.gcs.output.mode=<kafka output mode>
```
Comment (shashank-google) [inline java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md:171]: add example command to run the template with absolute values

File: java/src/main/java/com/google/cloud/dataproc/templates/util/ReadSchemaUtil.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    Storage storage = StorageOptions.getDefaultInstance().getService();
+
+
+    try {
+      Blob blob = storage.get(bucket, objectUrl);
+      String schemaSource = new String(blob.getContent());
+      schema = (StructType) DataType.fromJson(schemaSource);
+    }
+    catch (Throwable th){
```
Comment (shashank-google) [inline java/src/main/java/com/google/cloud/dataproc/templates/util/ReadSchemaUtil.java:45]: no try catch please

File: java/src/main/resources/template.properties [hunk 5]
```diff
@@ -207,4 +207,19 @@ kafka.bq.await.termination.timeout=420000
...
+kafka.bootstrap.servers=
+kafka.topic=
+kafka.starting.offset=
+kafka.schema.url=
+
+#Kafka To GCS
+kafka.gcs.output.location=;
+kafka.gcs.output.format=parquet
+kafka.gcs.output.mode=append
+kafka.gcs.await.termination.timeout=600000
```
Comment (shashank-google) [inline java/src/main/resources/template.properties:225]: what is unit of measurement here.
Please rename the property to reflect the same like kafka.gcs.await.termination.timeout.msec / .sec / .min etc

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+          KAFKA_GCS_OUTPUT_LOCATION,
+          KAFKA_BOOTSTRAP_SERVERS,
+          KAFKA_TOPIC);
+      throw new IllegalArgumentException(
+          "Required parameters for KafkaToGCS not passed. "
+              + "Set mandatory parameter for KafkaToGCS template "
+              + "in resources/conf/template.properties file.");
+    }
+
+    if (kafkaMessageFormat.equals("json") & StringUtils.isAllBlank(kafkaSchemaUrl)) {
```
Comment (shashank-google) [inline java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java:83]: separate validation into a new validateInput() method

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+      LOGGER.error("{} is a required parameter for JSON format messages", KAFKA_SCHEMA_URL);
+      throw new IllegalArgumentException("Required parameters for KafkaToGCS not passed.");
+    }
+
+    SparkSession spark = null;
+    LOGGER.info(
```
Comment (shashank-google) [inline java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java:89]: move logger into validate input method

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+        KAFKA_GCS_OUTPUT_MODE,
+        kafkaOutputMode,
+        KAFKA_GCS_AWAIT_TERMINATION_TIMEOUT,
+        kafkaAwaitTerminationTimeout);
+
+    try {
+      // Initialize Spark session
+      spark = SparkSession.builder().appName("Spark KafkaToGCS Job").getOrCreate();
+
+      LOGGER.debug("added jars : {}", spark.sparkContext().addedJars().keys());
```
Comment (shashank-google) [inline java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java:117]: do we need this logger ?
I am ok if we want to keep it, but highlighting if it there by mistake.

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+      KafkaReader reader = new KafkaReader();
+
+      LOGGER.info("Calling Kafka Reader");
+
+      // Reading Kafka stream into dataframe
+      Dataset<Row> processedData = reader.readKafkaTopic(spark, getProperties());
+
+      // Write the output to GCS location
+      processedData
```
Comment (shashank-google) [inline java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java:127]: Question:
will this code allow multiple threads (executors) to run in parallel ?
If so will each executor write independent file in gcs ?

**Review by shashank-google — CHANGES_REQUESTED** [review]

**Comment by shashank-google** [comment]
@nikhil6790 any updates on this one

**Comment by shashank-google** [comment]
Given no movement on this pull request. I am closing it and unassigning @nikhil6790 from it.

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java (hunk 2 above)
Comment (nikhil6790) [reply java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java]: Removed try catch blocks

File: java/src/main/java/com/google/cloud/dataproc/templates/util/ReadSchemaUtil.java (hunk 4 above)
Comment (nikhil6790) [reply java/src/main/java/com/google/cloud/dataproc/templates/util/ReadSchemaUtil.java]: removed try catch block

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md (hunk 3 above)
Comment (nikhil6790) [reply java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md:171]: Added absolute values to job submit command

File: java/src/main/resources/template.properties (hunk 5 above)
Comment (nikhil6790) [reply java/src/main/resources/template.properties]: Appended ".ms" to all instances of termination timeout to indicate time unit

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java (hunk 8 above)
Comment (nikhil6790) [reply java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java]: Removed the logger. Not required

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java (hunk 9 above)
Comment (nikhil6790) [reply java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java]: Spark handles spawning additional parallel executors to scale for multiple partitions or data volume

File: java/README.md [hunk 10]
```diff
@@ -19,6 +19,7 @@ Please refer to the [Dataproc Templates (Java - Spark) README](java/README.md)
 * [GCSToJDBC](/java/src/main/java/com/google/cloud/dataproc/templates/gcs/README.md) (blogpost [link](https://medium.com/google-cloud/importing-data-from-gcs-to-databases-via-jdbc-using-dataproc-serverless-7ed75eab93ba))
 * [SnowflakeToGCS](/java/src/main/java/com/google/cloud/dataproc/templates/snowflake/README.md)
 * [KafkaToBQ](/java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md) (blogpost [link](https://medium.com/google-cloud/export-data-from-apache-kafka-to-bigquery-using-dataproc-serverless-4a666535117c))
+* [KafkaToGCS](/java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md)
```
Comment (shashank-google) [inline java/README.md:22]: Same needs to be done in /README.md as well

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaReader.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+      case "json":
+        LOGGER.debug("Processing JSON Message");
+        processedData = processJsonMessage(inputData);
+        break;
+    }
+
+    return processedData;
+  }
+
+  private Dataset<Row> processJsonMessage(Dataset<Row> df) {
```
Comment (shashank-google) [inline java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaReader.java:81]: Can we please write a junit test case to unit test this method.

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaReader.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+        df.selectExpr("CAST(key AS STRING)", "CAST(value AS STRING)")
+            .select(
+                functions.col("key"),
+                functions.from_json(functions.col("value"), schema).alias("json_value"))
+            .select("key", "json_value.*");
+
+    return processedDF;
+  }
+
+  private Dataset<Row> processBytesMessage(Dataset<Row> df) {
```
Comment (shashank-google) [inline java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaReader.java:95]: Can we please write a junit test case to unit test this method.

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+          .start()
+          .awaitTermination(kafkaAwaitTerminationTimeout);
+
+      LOGGER.info("KakfaToGCS job completed.");
+      spark.stop();
+  }
+
+
+  void validateInputs(){
+    if (StringUtils.isAllBlank(gcsOutputLocation)
```
Comment (shashank-google) [inline java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java:99]: We are trying to improve codebase by using jakarta validatations (eg: JDBCToGCSConfig.java)
You can possibly make a followup github issue for it, if you want.

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md [hunk 14]
```diff
@@ -125,4 +125,69 @@ bin/start.sh \
...
+bin/start.sh \
+-- --template KAFKATOGCS \
+--templateProperty project.id=<gcp-project-id> \
+--templateProperty kafka.gcs.output.location=<gcs path> \
+--templateProperty kafka.bootstrap.servers=<kafka broker list> \
+--templateProperty kafka.topic=<kafka topic name> \
+```
+
+
+### Configurable Parameters
```
Comment (shashank-google) [inline java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md:150]: `Configurable Parameters` should be removed.
As it is just a duplication of template.properties file

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md [hunk 15]
```diff
@@ -125,4 +125,69 @@ bin/start.sh \
 --templateProperty kafka.bq.table=kafkaevents \
 --templateProperty kafka.bq.temp.gcs.bucket=templates-demo-kafkatobq \
 --templateProperty kafka.bq.await.termination.timeout=1200000
+```
+
+
+## 2. Kafka To GCS
```
Comment (shashank-google) [inline java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md:131]: You should specify the behaviour of this template as well.
Like how many kafka messages will be written in gcs output file.
What is buffer time window and buffer message limit to flush out these messages into a gcs file.
Will it update existing file or create a new file.
If windowed writes are not being done then we should create a new gitissue to enhance template to add the functionality.

File: java/src/main/java/com/google/cloud/dataproc/templates/util/ReadSchemaUtil.java [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+import com.google.cloud.storage.StorageOptions;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.StructType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ReadSchemaUtil {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ReadSchemaUtil.class);
+
+  public static StructType readSchema(String schemaUrl)  {
```
Comment (shashank-google) [inline java/src/main/java/com/google/cloud/dataproc/templates/util/ReadSchemaUtil.java:29]: add unit tests for same

File: java/src/main/resources/template.properties [hunk 17]
```diff
@@ -242,6 +242,20 @@ kafka.bq.fail.on.dataloss=false
...
+
+#Kafka Common
+kafka.message.format=
+kafka.bootstrap.servers=
+kafka.topic=
+kafka.starting.offset=
+kafka.schema.url=
+
+#Kafka To GCS
+kafka.gcs.output.location=;
```
Comment (shashank-google) [inline java/src/main/resources/template.properties:254]: why location has `;` ?

File: java/src/main/resources/template.properties [hunk 18]
```diff
@@ -242,6 +242,20 @@ kafka.bq.fail.on.dataloss=false
 ## Ouptut mode for writing data. Accepted values: 'append', 'complete', 'update'
 kafka.bq.stream.output.mode=append
 
+
+#Kafka Common
+kafka.message.format=
```
Comment (shashank-google) [inline java/src/main/resources/template.properties:247]: we need to follow consistent naming pattern for config variables like
`kafka.gcs.input.message.format`
And so on for all the inputs and outputs

**Review by shashank-google — CHANGES_REQUESTED** [review]

File: java/README.md (hunk 10 above)
Comment (nikhil6790) [reply java/README.md:22]: Changes done. Will push and resolve conversation.

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java (hunk 13 above)
Comment (nikhil6790) [reply java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java:99]: Will raise a follow up github issue to include jakarta validations

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md (hunk 14 above)
Comment (nikhil6790) [reply java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md:150]: Configuration parameters have been removed. Will push the changes and resolve conversation

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md (hunk 15 above)
Comment (nikhil6790) [reply java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md:131]: Template support default micro-batches.
I am introducing time based micro batching to better control output file generation.
I will be adding the detailed template behaviour post these additions

File: java/src/main/resources/template.properties (hunk 17 above)
Comment (nikhil6790) [reply java/src/main/resources/template.properties:254]: Typo fixed. Will push the changes and resolve the conversation

File: java/src/main/resources/template.properties (hunk 18 above)
Comment (nikhil6790) [reply java/src/main/resources/template.properties:284]: As discussed , the sink value is deliberately omitted as these are source related parameters and the reusable KafkaReader can be used for any Kafka based Template irrespective of Sink

File: java/src/main/java/com/google/cloud/dataproc/templates/util/ReadSchemaUtil.java (hunk 16 above)
Comment (nikhil6790) [reply java/src/main/java/com/google/cloud/dataproc/templates/util/ReadSchemaUtil.java:29]: Skipping Test for Util class that reads GCS bucket. Require explicit GCS object creation to test functionality

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md (hunk 15 above)
Comment (nikhil6790) [reply java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md:131]: Micro batching  configurations will be introduced in followup git issue.

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaReader.java (hunk 12 above)
Comment (nikhil6790) [reply java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaReader.java]: Added JUnit test case
https://github.com/GoogleCloudPlatform/dataproc-templates/commit/f431013548ff3c3c0d28a293d6dd81279f091ce9

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaReader.java (hunk 11 above)
Comment (nikhil6790) [reply java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaReader.java]: Added JUnit test case
https://github.com/GoogleCloudPlatform/dataproc-templates/commit/f431013548ff3c3c0d28a293d6dd81279f091ce9

**Comment by shashank-google** [comment]
@nikhil6790 thanks for resolving all the comments. In general it looks good to me.
@hhasija or @vanshaj-bhatia  Can you please test, review and merge this pull request.

**Review by shashank-google — APPROVED** [review]

**Pull request merged**